src/dbnode/storage/index.go

```diff
@@ -716,77 +723,196 @@ func (i *nsIndex) Query(
 	opts index.QueryOptions,
 ) (index.QueryResults, error) {
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		return index.QueryResults{}, errDbIndexUnableToQueryClosed
 	}
 
-	// override query response limit if needed.
+	// Track this as an inflight query that needs to finish
+	// when the index is closed
+	i.queriesWg.Add(1)
+	defer i.queriesWg.Done()
+
+	// Enact overrides for query options
+	opts = i.overriddenOptsForQueryWithLock(opts)
+	timeout := i.timeoutForQueryWithLock(ctx)
+
+	// Retrieve blocks to query, then we can release lock
+	// NB(r): Important not to block ticking, and other tasks by
+	// holding the RLock during a query
+	blocks, err := i.blocksForQueryWithLock(xtime.NewRanges(xtime.Range{
+		Start: opts.StartInclusive,
+		End:   opts.EndExclusive,
+	}))
+
+	// Can now release the lock and execute the query without holding the lock
+	i.state.RUnlock()
+
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	var (
+		start      = i.nowFn()
+		deadline   = start.Add(timeout)
+		exhaustive = true
+		results    = i.newConcurrentResults(ctx)
+		multiErr   = xerrors.NewMultiError()
+		workers    = i.opts.QueryIDsWorkerPool()
+		wg         sync.WaitGroup
+	)
+	for _, block := range blocks {
+		// Capture block for async query execution below
+		block := block
+
+		// terminate early if we know we don't need any more results
+		if opts.Limit > 0 && results.Size() >= opts.Limit {
+			results.Lock()
+			exhaustive = false
+			results.Unlock()
+			break
+		}
+
+		execBlockQuery := func() {
```

Comment:
The variable capture above and allocating this closure each time isn't required. I think you can reduce it all down to a single closure allocation by moving the function declaration above the loop and then changing the definition of the function to be:

```golang
func(block index.Block) {
    ...
}
```

and then down below it becomes

```golang
workers.Go(func() {
    execBlockQuery(block) // Capture happens for free here
    wg.Done()
})
```

Reply:
Should reduce number of allocated functions per block from 2->1